add victoria-metrics-distributed chart (#960)
VictoriaMetrics/helm-charts: charts/victoria-metrics-distributed/README.md

[![Slack](https://img.shields.io/badge/join%20slack-%23victoriametrics-brightgreen.svg)](https://slack.victoriametrics.com/)

A Helm chart for Running VMCluster on Multiple Availability Zones

## Prerequisites

* Install the follow packages: ``git``, ``kubectl``, ``helm``, ``helm-docs``. See this [tutorial](../../REQUIREMENTS.md).

* PV support on underlying infrastructure.

* Multiple availability zones.

## Chart Details

This chart sets up multiple VictoriaMetrics cluster instances on multiple [availability zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/), provides both global write and read entrypoints.

The default setup is as shown below:

![victoriametrics-distributed-topology](./victoriametrics-distributed-topology.png)

For write:
1. extra-vmagent(optional): scrapes external targets and all the components installed by this chart, sends data to global write entrypoint.
2. vmauth-global-write: global write entrypoint, proxies requests to one of the zone `vmagent` with `least_loaded` policy.

> 3. vmagent(per-zone): remote writes data to availability zones that enabled `.Values.availabilityZones.allowIngest`.

Reviewer: we need to mention that it acts as a buffer in the same time, for restoring data consistency in case if any zone is unavailable. Hence, we need to provide recommendations regarding disk space capacity.

VictoriaMetrics supports [data replication](https://docs.victoriametrics.com/cluster-victoriametrics/#replication-and-data-safety) natively which can guarantees data availability when part of the vmstorage instances failed. But it doesn't works well if vmstorage instances are spread on multiple availability zones, since data replication could be stored on single availability zone, which will be lost when zone outage happens.    
To avoid this, vmcluster must be installed on multiple availability zones and each of them should contain complete data. In this way, as long as one zone is available, both global write and read entrypoint work without interruption.

### How to write data?

The chart provides `vmauth-global-write` as global write entrypoint, it supports [push-based data ingestion protocols](https://docs.victoriametrics.com/vmagent/#how-to-push-data-to-vmagent) as VictoriaMetrics does.
You can use this entrypoint as vmagent's remoteWrite url or use other clients push data to it directly.

### How to query data?

> The chart provides `vmauth-global-read` as global read entrypoint, it picks the first zone as it's preferred datasource and switches automatically to next zone if first one is unavailable, check [vmauth `first_available`](https://docs.victoriametrics.com/vmauth/#high-availability) for more details.  

Reviewer: ```suggestion
The chart provides `vmauth-global-read` as global read entrypoint, it picks the first available zone (see [first_available](https://docs.victoriametrics.com/vmauth/#high-availability) policy) as it's preferred datasource and switches automatically to next zone if first one is unavailable.  
```

> If you have workloads like vmalert or grafana which deployed on one zone specifically, you can use `vmauth-read-proxy` on the same zone as your datasource, since it always prefer "local" vmcluster which reduces cross-zone traffic. 

Reviewer: ```suggestion
If you have services like [vmalert](https://docs.victoriametrics.com/vmalert.html) or Grafana deployed in each zone, then configure them to use local `vmauth-read-proxy`. Per-zone `vmauth-read-proxy` always prefers "local" vmcluster for querying and reduces cross-zone traffic. 
```

You can also pick other proxies like kubernetes service which supports [Topology Aware Routing](https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/) as global read entrypoint.

> ### What will happen if one zone outraged?
> 
> If availability zone `zone-eu-1` outraged, `vmauth-global-write` and `vmauth-global-read` still work without interruption, and following things happen:

Reviewer 2: ```suggestion
### What will happen in case of zone outage?

In case availability zone `zone-eu-1` is not available, `vmauth-global-write` and `vmauth-global-read` still work without interruption, and following things happen:
```

Reviewer 2: Most likely "outrage" was used due to a typo :)

Author: yes, thanks!

1. `vmauth-global-write` stops proxying write requests to `zone-eu-1` automatically;
2. `vmauth-global-read` and `vmauth-read-proxy` stops proxying read requests to `zone-eu-1` automatically;

> 3. `vmagent` on `zone-us-1` fails to send data to `zone-eu-1` `vmauth-write-balancer`, starts to [buffer data on disk](https://docs.victoriametrics.com/vmagent/#calculating-disk-space-for-persistence-queue) if `-remoteWrite.disableOnDiskQueue` is not specified.

Reviewer: ```suggestion
3. `vmagent` on `zone-us-1` fails to send data to `zone-eu-1.vmauth-write-balancer`, starts to [buffer data on disk](https://docs.victoriametrics.com/vmagent/#calculating-disk-space-for-persistence-queue) (unless `-remoteWrite.disableOnDiskQueue` is specified, which is not recommended for this topology).
```

> 

Reviewer: Once the availability of `zone-eu-1` is restored, `vmagent` from `zone-us-1` will re-send all the buffered data to `zone-eu-1` to restore data consistency.

Reviewer: @Haleygo @zekker6 when will vmauth start to route requests to `zone-eu-1`? Could it be that after the outage vmauth will route queries to `zone-eu-1` and will serve incomplete responses?

Reviewer 2: Based on [this](https://github.com/VictoriaMetrics/helm-charts/blob/64b4f71f94139548de4147c06a0907983901aaab/charts/victoria-metrics-distributed/victoriametrics-distributed-topology.png) diagram it will start routing requests after vmauth read proxy will become available. So in this case it is possible that it will deliver incomplete responses.

Author: Add more description for it, plz check https://github.com/VictoriaMetrics/helm-charts/pull/960/commits/19f29bedc9a77a5c285cc84bfdca32ef3f8ac80c

# Victoria Metrics Helm Chart for Setting up VMCluster on Multiple Availability Zones

![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square)  ![Version: 0.1.0](https://img.shields.io/badge/Version-0.1.0-informational?style=flat-square)
[![Artifact Hub](https://img.shields.io/endpoint?url=https://artifacthub.io/badge/repository/victoriametrics)](https://artifacthub.io/packages/helm/victoriametrics/victoria-metrics-distributed)
[![Slack](https://img.shields.io/badge/join%20slack-%23victoriametrics-brightgreen.svg)](https://slack.victoriametrics.com/)

A Helm chart for Running VMCluster on Multiple Availability Zones

## Prerequisites

* Install the follow packages: ``git``, ``kubectl``, ``helm``, ``helm-docs``. See this [tutorial](../../REQUIREMENTS.md).

* PV support on underlying infrastructure.

* Multiple availability zones.

> ## Chart Details

Reviewer 3: 1. I think we should place scheme from issue here
2. Maybe we should describe upgrade scenario for the cluster step-by-step

Author: added in https://github.com/VictoriaMetrics/helm-charts/pull/960/commits/4e3204ed23a119eb48b9a2eb6d9a8ce8f84156a8, please check.

# Victoria Metrics Helm Chart for Setting up VMCluster on Multiple Availability Zones

![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square)  ![Version: 0.1.0](https://img.shields.io/badge/Version-0.1.0-informational?style=flat-square)
[![Artifact Hub](https://img.shields.io/endpoint?url=https://artifacthub.io/badge/repository/victoriametrics)](https://artifacthub.io/packages/helm/victoriametrics/victoria-metrics-distributed)
[![Slack](https://img.shields.io/badge/join%20slack-%23victoriametrics-brightgreen.svg)](https://slack.victoriametrics.com/)

A Helm chart for Running VMCluster on Multiple Availability Zones

## Prerequisites

* Install the follow packages: ``git``, ``kubectl``, ``helm``, ``helm-docs``. See this [tutorial](../../REQUIREMENTS.md).

* PV support on underlying infrastructure.

* Multiple availability zones.

> ## Chart Details

Reviewer: Thanks for writing the doc! But I find it lacking a bit of details on the topology. Maybe explaining failure scenarios and how this topology helps to provide availability will be more understandable for reader.

This chart is for setting up multiple VictoriaMetrics cluster instances on multiple [availability zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/).

> The default values suppose there are two availability zones with label `topology.kubernetes.io/zone: zone-a` and `topology.kubernetes.io/zone: zone-a`, default topology will be like below:

Reviewer: Should it be zone-a and zone-b?

Author: fixed!

![victoriametrics-distributed-topology](./victoriametrics-distributed-topology.jpg)

> It contains required availability zone components and global components, and other optional resources.

Reviewer: This sentence doesn't contribute additional knowledge and can be skipped.

Author: removed

> Availability zone components:

Reviewer: ```suggestion
Availability zone components:
```
Each AZ contains the following components:

Author: rephrased

1. vmagent: remotes write to all availability zones to ensure data completeness.
2. vmauth-ingest: performs `least_loaded` policy for spreading requests to vminsert instances located in the same availability zone.
3. vmcluster: supports different setup per availability zone.
4. vmauth-query: performs `least_loaded` policy for spreading requests to vmselect instances located in the same availability zone.

> 5. vmauth-query-cross-az: performs `first_available` policy on vmselect instances of all the availability zones, prefers "local" vmselect to reduce cross zone network traffic.

Reviewer: Can it be substituted with ` vmauth-global-query`?

Author: changed some namse, plz check the latest and let me know)

### When&Why use `victoria-metrics-distributed` chart?

One of the best practice of running production kubernetes cluster is running with [multiple availability zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/). And apart from kubernetes control plane components, we also want to spread our application pods on multiple zones to continue working during zone outage.
There are two ways to set up VictoriaMetrics cluster on multiple availability zones:
1. using one vmcluster but spread the components on different zones by `topologySpreadConstraints`.
For example, set up vmcluster with `replicas: 2` for vminsert, vmstorage and vmselect with `replicationFactor=2`, and set `topologySpreadConstraints` for all of the components:
```
    topologySpreadConstraints:
      - maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: DoNotSchedule
```
Each zone will have a pair of nodes of each type - vminsert, vmselect and vmstorage, thus vmcluster can still serve both write and query when one zone is compeletely down.

> But VictoriaMetrics components themselves don't support zone-awareness, means in the above example, if you have three vmstorage nodes with `replicationFactor=2`, it can't be ensure that duplicated data are written to vmstorage instances on different availability zone, thus data could be incomplete when one availability zone is down.

Reviewer: Doesn't this mean this approach is incorrect?) If yes, we shouldn't mention it.

Author: yes, removed

```console
helm list -f vmcluster -n NAMESPACE
```

See the history of versions of ``vmcluster`` application with command.

```console
helm history vmcluster -n NAMESPACE
```

## How to uninstall

Remove application with command.

```console
helm uninstall vmcluster -n NAMESPACE
```

## Documentation of Helm Chart

Install ``helm-docs`` following the instructions on this [tutorial](../../REQUIREMENTS.md).

Generate docs with ``helm-docs`` command.

```bash
cd charts/victoria-metrics-distributed`

helm-docs
```

The markdown generation is entirely go template driven. The tool parses metadata from charts and generates a number of sub-templates that can be referenced in a template file (by default ``README.md.gotmpl``). If no template file is provided, the tool has a default internal template that will generate a reasonably formatted README.

## Parameters

The following tables lists the configurable parameters of the chart and their default values.

Change the values according to the need of the environment in ``victoria-metrics-distributed`/values.yaml`` file.

| Key | Type | Default | Description |
|-----|------|---------|-------------|

> | availabilityZones[0].allowIngest | bool | `true` |  |

Reviewer: Is it planned to add description to these params at some point?

Author: they were there, but with wrong comment format, fixed.
